### ISSUE TITLE
Does not handle fgets returning an error gracefully

### ISSUE THREAD

**Issue by mglaman** [TYPO3/phar-stream-wrapper#68]
If `fgets` returns `false` due to an error, this package will crash as it passes a `bool` to `strpos`:

```
TypeError: strpos() expects parameter 1 to be string, bool given in strpos() (line 109 of /var/www/html/jwtest-com/vendor/typo3/phar-stream-wrapper/src/Phar/Reader.php).
```

I'm still not certain _how_ we got an error from `fgets`. We were trying to launch the phpstan.phar:

```
TYPO3\PharStreamWrapper\Phar\Reader->extractData('/var/www/html/jwtest-com/vendor/phpstan/phpstan/phpstan.phar') (Line: 53)
```

Relevant issue on Drupal.org. This gets triggered with:

```
    if (!class_exists('PHPStan\ExtensionInstaller\GeneratedConfig')) {
```

That causes the PHPStan\PharAutoloader to execute:

```
PHPStan\PharAutoloader::loadClass('PHPStan\ExtensionInstaller\GeneratedConfig')
spl_autoload_call('PHPStan\ExtensionInstaller\GeneratedConfig')
class_exists('PHPStan\ExtensionInstaller\GeneratedConfig') (Line: 589)
```

And we end up broken. I don't know how many lines it takes to get an error, but it's before `feof` returns true.

**Comment by ohader** [comment]
Thanks for reporting this. Is there a ticket in Drupal's issue tracker? I'm just trying to get some more scenario and context details.
Anyway, I'll try to reproduce that behavior soonish.

**Comment by ohader** [comment]
Okay... independent from PharStreamWrapper I could reproduce `fgets` resolving to `false` at the end of a file offset, without(!) having `feof() === true`. Basically I was reading 2-byte chunks from a total of 13 bytes. I was able to reproduce that on macOS with PHP 7.4, but not with PHP 8.0. Well...

I think it's fine to add an additional guard for this case. Still, finding out what really happens with `phpstan.phar` of course would be interesting.

**Comment by ohader** [comment]
@mglaman Please let me know whether PR #70 fixes the issue for Drupal. Thx

**Comment by mglaman** [comment]
Sorry! I linked this GitHub issue to the Drupal issue, but not here to there. Drupal issue: https://www.drupal.org/project/upgrade_status/issues/3232011

**Comment by mglaman** [comment]
> I was able to reproduce that on macOS with PHP 7.4, but not with PHP 8.0. Well...

I had PHP 8.0, locally maybe that's why I couldn't reproduce. I'll see if I can work the user to get a reproduction and more data.

**Comment by ohader** [comment]
Let's have a look, when that actually happens - thus, when PHP has problems to read from a stream, without issuing `feof`.

* https://github.com/php/php-src/blob/PHP-7.4.20/ext/standard/file.c#L1122-L1127

`false` is returned when underlying stream invocation returns `null`, let's have a look...

* https://github.com/php/php-src/blob/PHP-7.4.20/main/streams/streams.c#L991-L996

`null` is returned nothing could be read from a stream (zero bytes copied to memory).

**tl;dr:** `fgets($resource)` returns `false`, when not a single byte could be read from a string, as indicated by variable `total_copied`.

Was was not able to reproduce this behavior with DDEV and PHP 7.4.20 on macOS.
